In [6]:
!pip uninstall -y bitsandbytes

!pip install -q -U \
    transformers \
    accelerate \
    peft \
    datasets \
    bitsandbytes \
    sentencepiece \
    trl

Found existing installation: bitsandbytes 0.50.2
Uninstalling bitsandbytes-0.50.2:
  Successfully uninstalled bitsandbytes-0.50.2


In [7]:
import torch
import bitsandbytes as bnb

print("PyTorch:", torch.__version__)
print("PyTorch CUDA:", torch.version.cuda)
print("CUDA available:", torch.cuda.is_available())
print("GPU:", torch.cuda.get_device_name(0))
print("bitsandbytes:", bnb.__version__)
print("bitsandbytes path:", bnb.__file__)

PyTorch: 2.6.0+cu124
PyTorch CUDA: 12.4
CUDA available: True
GPU: Tesla T4
bitsandbytes: 0.50.2
bitsandbytes path: /usr/local/lib/python3.13/dist-packages/bitsandbytes/__init__.py


In [8]:
import torch

print("PyTorch version:", torch.__version__)
print("GPU available:", torch.cuda.is_available())

if torch.cuda.is_available():
    print("GPU:", torch.cuda.get_device_name(0))

PyTorch version: 2.6.0+cu124
GPU available: True
GPU: Tesla T4


In [9]:
import bitsandbytes as bnb

print("bitsandbytes version:", bnb.__version__)

test = torch.tensor(
    [[1.0, 2.0], [3.0, 4.0]],
    device="cuda",
    dtype=torch.float16
)

print("CUDA tensor test:", test.device)

bitsandbytes version: 0.50.2
CUDA tensor test: cuda:0


In [10]:
import transformers
import accelerate
import bitsandbytes
import sentencepiece

print("transformers:", transformers.__version__)
print("accelerate:", accelerate.__version__)
print("bitsandbytes:", bitsandbytes.__version__)
print("sentencepiece:", sentencepiece.__version__)

transformers: 5.17.0
accelerate: 1.15.0
bitsandbytes: 0.50.2
sentencepiece: 0.2.2


In [11]:
import torch
import bitsandbytes as bnb

print("PyTorch:", torch.__version__)
print("PyTorch CUDA:", torch.version.cuda)
print("CUDA available:", torch.cuda.is_available())

if torch.cuda.is_available():
    print("GPU:", torch.cuda.get_device_name(0))

print("bitsandbytes:", bnb.__version__)

PyTorch: 2.6.0+cu124
PyTorch CUDA: 12.4
CUDA available: True
GPU: Tesla T4
bitsandbytes: 0.50.2


In [12]:
!python -m bitsandbytes

=================== bitsandbytes v0.50.2 ===================
Platform: Linux-6.6.122+-x86_64-with-glibc2.39
  libc: glibc-2.39
Python: 3.13.15
PyTorch: 2.6.0+cu124
  CUDA: 12.4
  HIP: N/A
  XPU: N/A
Related packages:
  accelerate: 1.15.0
  diffusers: 0.40.0
  numpy: 2.5.3
  pip: 24.1.2
  peft: 0.21.0
  safetensors: 0.8.0
  transformers: 5.17.0
  triton: 3.2.0
  trl: 1.14.0
PyTorch settings found: CUDA_VERSION=124, Highest Compute Capability: (7, 5).
Checking that the library is importable and callable...
SUCCESS!


In [13]:
from transformers import AutoTokenizer, AutoModelForCausalLM
import torch

model_name = "Qwen/Qwen2.5-1.5B-Instruct"

tokenizer = AutoTokenizer.from_pretrained(model_name)

model = AutoModelForCausalLM.from_pretrained(
    model_name,
    torch_dtype=torch.float16,
    device_map="auto"
)

print("Model loaded successfully!")

[transformers] `torch_dtype` is deprecated! Use `dtype` instead!


Loading weights:   0%|          | 0/338 [00:00<?, ?it/s]

Model loaded successfully!


In [14]:
messages = [
    {
        "role": "user",
        "content": "Explain what a business KPI is in two simple sentences."
    }
]

text = tokenizer.apply_chat_template(
    messages,
    tokenize=False,
    add_generation_prompt=True
)

inputs = tokenizer(
    text,
    return_tensors="pt"
).to(model.device)

outputs = model.generate(
    **inputs,
    max_new_tokens=80,
    do_sample=False
)

response = tokenizer.decode(
    outputs[0][inputs["input_ids"].shape[1]:],
    skip_special_tokens=True
)

print(response)

A Business Key Performance Indicator (KPI) is a measurable value that represents the performance of an organization or its key processes and helps to evaluate progress towards strategic goals.


In [15]:
import json

insights = {
    "dataset": {
        "name": "sales.csv",
        "domain": "sales",
        "record_count": 12500,
        "columns": {
            "numeric": [
                "sales",
                "profit",
                "quantity"
            ],
            "categorical": [
                "product",
                "category",
                "region"
            ],
            "temporal": [
                "date"
            ],
            "identifier": [
                "order_id"
            ]
        }
    },

    "metrics": [
        {
            "name": "Total Revenue",
            "value": 1250000,
            "unit": "INR",
            "type": "sum",
            "description": "Total revenue across all records"
        },
        {
            "name": "Profit Margin",
            "value": 19.39,
            "unit": "percent",
            "type": "percentage",
            "description": "Profit as a percentage of revenue"
        }
    ],

    "trends": [
        {
            "metric": "Revenue",
            "dimension": "date",
            "direction": "increasing",
            "change": 12.4,
            "unit": "percent",
            "period": {
                "start": "2026-01-01",
                "end": "2026-03-31"
            },
            "significance": "high"
        }
    ],

    "comparisons": [
        {
            "dimension": "category",
            "metric": "Revenue",
            "highest": {
                "name": "Electronics",
                "value": 450000
            },
            "lowest": {
                "name": "Furniture",
                "value": 180000
            }
        }
    ],

    "anomalies": [],

    "distributions": [],

    "visualizations": [
        {
            "title": "Monthly Revenue",
            "type": "line",
            "metric": "Revenue",
            "dimension": "date",
            "summary": "Revenue increased consistently during the reporting period."
        }
    ]
}
insights_json = json.dumps(insights, indent=2)
print(insights_json)

{
  "dataset": {
    "name": "sales.csv",
    "domain": "sales",
    "record_count": 12500,
    "columns": {
      "numeric": [
        "sales",
        "profit",
        "quantity"
      ],
      "categorical": [
        "product",
        "category",
        "region"
      ],
      "temporal": [
        "date"
      ],
      "identifier": [
        "order_id"
      ]
    }
  },
  "metrics": [
    {
      "name": "Total Revenue",
      "value": 1250000,
      "unit": "INR",
      "type": "sum",
      "description": "Total revenue across all records"
    },
    {
      "name": "Profit Margin",
      "value": 19.39,
      "unit": "percent",
      "type": "percentage",
      "description": "Profit as a percentage of revenue"
    }
  ],
  "trends": [
    {
      "metric": "Revenue",
      "dimension": "date",
      "direction": "increasing",
      "change": 12.4,
      "unit": "percent",
      "period": {
        "start": "2026-01-01",
        "end": "2026-03-31"
      },
      "significa

In [16]:
prompt = f"""
You are a business analytics assistant.

Generate a concise analytical narrative from the structured
analytical findings provided below.

Rules:
1. Use ONLY the information provided.
2. Do not invent numbers, facts, trends, or causes.
3. Mention the most important metrics.
4. Mention important trends.
5. Mention important comparisons.
6. Mention anomalies if they are provided.
7. Ignore empty sections.
8. Do not claim that one factor caused another unless causation
   is explicitly provided.
9. Write in a professional business-analysis style.

Structured analytical findings:

{insights_json}

Generate the business narrative now.
"""

In [17]:
messages = [
    {
        "role": "user",
        "content": prompt
    }
]

text = tokenizer.apply_chat_template(
    messages,
    tokenize=False,
    add_generation_prompt=True
)

inputs = tokenizer(
    text,
    return_tensors="pt"
).to(model.device)

outputs = model.generate(
    **inputs,
    max_new_tokens=300,
    temperature=0.2,
    do_sample=True
)

response = tokenizer.decode(
    outputs[0][inputs["input_ids"].shape[1]:],
    skip_special_tokens=True
)

print(response)

The dataset contains sales data for 12,500 orders spanning from January to March 2026. The total revenue generated was INR 1,250,000, with a profit margin of 19.39%. This indicates that for every rupee spent on sales, approximately 19.39 cents were turned into profit.

One significant trend observed is an increasing revenue over time. Specifically, the revenue has shown a 12.4% increase between January and March 2026 compared to the same period in 2025. This upward trajectory suggests strong growth in sales activity during this quarter.

In terms of product performance, Electronics products have been particularly successful, accounting for 45% of total revenue. In contrast, Furniture products contribute only 18%, indicating a clear preference among customers for electronic goods.

Comparatively, the highest revenue category is Electronics, generating INR 450,000, while the lowest is Furniture at INR 180,000. These figures highlight the diverse nature of customer preferences within the 

In [18]:
healthcare_insights = {
    "dataset": {
        "name": "patients.csv",
        "domain": "healthcare",
        "record_count": 5000,
        "columns": {
            "numeric": [
                "age",
                "wait_time"
            ],
            "categorical": [
                "department",
                "diagnosis"
            ],
            "temporal": [],
            "identifier": [
                "patient_id"
            ]
        }
    },

    "metrics": [
        {
            "name": "Average Wait Time",
            "value": 42.5,
            "unit": "minutes",
            "type": "average",
            "description": "Average patient waiting time"
        },
        {
            "name": "Readmission Rate",
            "value": 8.2,
            "unit": "percent",
            "type": "percentage",
            "description": "Percentage of patients readmitted"
        }
    ],

    "trends": [],

    "comparisons": [
        {
            "dimension": "department",
            "metric": "wait_time",
            "highest": {
                "name": "Emergency",
                "value": 68.4
            },
            "lowest": {
                "name": "Dermatology",
                "value": 21.7
            }
        }
    ],

    "anomalies": [],
    "distributions": [],

    "visualizations": [
        {
            "title": "Average Wait Time by Department",
            "type": "bar",
            "metric": "wait_time",
            "dimension": "department",
            "summary": "Emergency has the highest average wait time."
        }
    ]
}
healthcare_json = json.dumps(
    healthcare_insights,
    indent=2
)

In [21]:
healthcare_json = json.dumps(
    healthcare_insights,
    indent=2
)

prompt = f"""
You are a business and analytics assistant.

Generate a concise analytical narrative from the structured
analytical findings provided below.

Rules:
1. Use ONLY the information provided.
2. Do not invent numbers, facts, trends, or causes.
3. Mention the most important metrics.
4. Mention important comparisons.
5. Mention trends only if they are provided.
6. Mention anomalies only if they are provided.
7. Ignore empty sections.
8. Do not claim that one factor caused another unless causation
   is explicitly provided.
9. Write in a professional and clear analytical style.
10. Adapt the terminology to the dataset's domain.

Structured analytical findings:

{healthcare_json}

Generate the analytical narrative now.
"""

print(prompt)



You are a business and analytics assistant.

Generate a concise analytical narrative from the structured
analytical findings provided below.

Rules:
1. Use ONLY the information provided.
2. Do not invent numbers, facts, trends, or causes.
3. Mention the most important metrics.
4. Mention important comparisons.
5. Mention trends only if they are provided.
6. Mention anomalies only if they are provided.
7. Ignore empty sections.
8. Do not claim that one factor caused another unless causation
   is explicitly provided.
9. Write in a professional and clear analytical style.
10. Adapt the terminology to the dataset's domain.

Structured analytical findings:

{
  "dataset": {
    "name": "patients.csv",
    "domain": "healthcare",
    "record_count": 5000,
    "columns": {
      "numeric": [
        "age",
        "wait_time"
      ],
      "categorical": [
        "department",
        "diagnosis"
      ],
      "temporal": [],
      "identifier": [
        "patient_id"
      ]
    }
  },


In [22]:
messages = [
    {
        "role": "user",
        "content": prompt
    }
]

text = tokenizer.apply_chat_template(
    messages,
    tokenize=False,
    add_generation_prompt=True
)

inputs = tokenizer(
    text,
    return_tensors="pt"
).to(model.device)

outputs = model.generate(
    **inputs,
    max_new_tokens=300,
    do_sample=False
)

response = tokenizer.decode(
    outputs[0][inputs["input_ids"].shape[1]:],
    skip_special_tokens=True
)

print(response)

The healthcare dataset contains records for 5,000 patients across various departments and diagnoses. The primary focus of this analysis is on two key metrics: the average wait time and the readmission rate.

The average wait time for all patients is 42.5 minutes, with an average readmission rate of 8.2 percent. These figures provide a baseline understanding of typical patient experiences within each department.

By examining the data further, we can identify specific areas where services may be more efficient or where improvements could be made. For instance, Emergency departments have the longest average wait times at 68.4 minutes compared to other departments like Dermatology, which has the shortest average wait time at 21.7 minutes.

These insights suggest that there might be opportunities for optimization in departments with longer wait times, such as Emergency, to reduce patient wait times and improve overall service efficiency. Additionally, focusing on reducing the readmission r

In [23]:
def validate_narrative(narrative, insights):
    """
    Basic validation to check whether important
    numerical values from the structured insights
    appear in the generated narrative.
    """

    important_values = []

    for metric in insights.get("metrics", []):
        important_values.append(str(metric["value"]))

    for trend in insights.get("trends", []):
        important_values.append(str(trend["change"]))

    for comparison in insights.get("comparisons", []):
        important_values.append(str(comparison["highest"]["value"]))
        important_values.append(str(comparison["lowest"]["value"]))

    missing = []

    for value in important_values:
        if value not in narrative:
            missing.append(value)

    return {
        "valid": len(missing) == 0,
        "missing_values": missing
    }

In [24]:
validation = validate_narrative(
    response,
    healthcare_insights
)

print(validation)

{'valid': True, 'missing_values': []}


In [25]:
training_data = [
    {
        "instruction": "Generate a concise analytical narrative from the structured findings. Use only the provided facts.",
        "input": {
            "domain": "sales",
            "metrics": [
                {
                    "name": "Total Revenue",
                    "value": 1250000,
                    "unit": "INR"
                },
                {
                    "name": "Profit Margin",
                    "value": 19.39,
                    "unit": "percent"
                }
            ],
            "trends": [
                {
                    "metric": "Revenue",
                    "direction": "increasing",
                    "change": 12.4,
                    "unit": "percent"
                }
            ],
            "comparisons": [
                {
                    "dimension": "category",
                    "metric": "Revenue",
                    "highest": {
                        "name": "Electronics",
                        "value": 450000
                    },
                    "lowest": {
                        "name": "Furniture",
                        "value": 180000
                    }
                }
            ],
            "anomalies": []
        },
        "output": "Revenue reached INR 1,250,000 with a profit margin of 19.39%. Revenue increased by 12.4% during the reporting period. Electronics generated the highest category revenue at INR 450,000, while Furniture generated the lowest at INR 180,000."
    },

    {
        "instruction": "Generate a concise analytical narrative from the structured findings. Use only the provided facts.",
        "input": {
            "domain": "healthcare",
            "metrics": [
                {
                    "name": "Average Wait Time",
                    "value": 42.5,
                    "unit": "minutes"
                },
                {
                    "name": "Readmission Rate",
                    "value": 8.2,
                    "unit": "percent"
                }
            ],
            "trends": [],
            "comparisons": [
                {
                    "dimension": "department",
                    "metric": "wait_time",
                    "highest": {
                        "name": "Emergency",
                        "value": 68.4
                    },
                    "lowest": {
                        "name": "Dermatology",
                        "value": 21.7
                    }
                }
            ],
            "anomalies": []
        },
        "output": "The average patient wait time was 42.5 minutes and the readmission rate was 8.2%. Emergency had the highest average wait time at 68.4 minutes, while Dermatology had the lowest at 21.7 minutes."
    },

    {
        "instruction": "Generate a concise analytical narrative from the structured findings. Use only the provided facts.",
        "input": {
            "domain": "sales",
            "metrics": [
                {
                    "name": "Total Orders",
                    "value": 8200,
                    "unit": "orders"
                },
                {
                    "name": "Average Order Value",
                    "value": 2450.50,
                    "unit": "INR"
                }
            ],
            "trends": [
                {
                    "metric": "Orders",
                    "direction": "decreasing",
                    "change": 6.5,
                    "unit": "percent"
                }
            ],
            "comparisons": [],
            "anomalies": []
        },
        "output": "The dataset contains 8,200 orders, with an average order value of INR 2,450.50. The number of orders decreased by 6.5% during the reporting period."
    },

    {
        "instruction": "Generate a concise analytical narrative from the structured findings. Use only the provided facts.",
        "input": {
            "domain": "retail",
            "metrics": [
                {
                    "name": "Total Sales",
                    "value": 850000,
                    "unit": "INR"
                }
            ],
            "trends": [],
            "comparisons": [],
            "anomalies": [
                {
                    "metric": "Sales",
                    "dimension": "date",
                    "observation": "2026-02-15",
                    "value": 95000,
                    "expected_range": {
                        "lower": 40000,
                        "upper": 65000
                    },
                    "type": "high",
                    "severity": "medium"
                }
            ]
        },
        "output": "Total sales were INR 850,000. A medium-severity high sales anomaly was detected on February 15, 2026, when sales reached INR 95,000 compared with the expected range of INR 40,000 to INR 65,000."
    },

    {
        "instruction": "Generate a concise analytical narrative from the structured findings. Use only the provided facts.",
        "input": {
            "domain": "healthcare",
            "metrics": [
                {
                    "name": "Average Treatment Cost",
                    "value": 5800,
                    "unit": "INR"
                },
                {
                    "name": "Patient Count",
                    "value": 3200,
                    "unit": "patients"
                }
            ],
            "trends": [],
            "comparisons": [],
            "anomalies": []
        },
        "output": "The dataset contains 3,200 patients, with an average treatment cost of INR 5,800. No trends, comparisons, or anomalies were identified in the provided analytical findings."
    }
]

In [26]:
import json

with open("training_data.jsonl", "w") as f:
    for example in training_data:
        f.write(json.dumps(example) + "\n")

print("Training examples:", len(training_data))
print("File created successfully.")

Training examples: 5
File created successfully.


In [27]:
with open("training_data.jsonl", "r") as f:
    print(f.read())

{"instruction": "Generate a concise analytical narrative from the structured findings. Use only the provided facts.", "input": {"domain": "sales", "metrics": [{"name": "Total Revenue", "value": 1250000, "unit": "INR"}, {"name": "Profit Margin", "value": 19.39, "unit": "percent"}], "trends": [{"metric": "Revenue", "direction": "increasing", "change": 12.4, "unit": "percent"}], "comparisons": [{"dimension": "category", "metric": "Revenue", "highest": {"name": "Electronics", "value": 450000}, "lowest": {"name": "Furniture", "value": 180000}}], "anomalies": []}, "output": "Revenue reached INR 1,250,000 with a profit margin of 19.39%. Revenue increased by 12.4% during the reporting period. Electronics generated the highest category revenue at INR 450,000, while Furniture generated the lowest at INR 180,000."}
{"instruction": "Generate a concise analytical narrative from the structured findings. Use only the provided facts.", "input": {"domain": "healthcare", "metrics": [{"name": "Average Wa

In [28]:
from datasets import load_dataset

dataset = load_dataset(
    "json",
    data_files="training_data.jsonl"
)

print(dataset)

Generating train split: 0 examples [00:00, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['instruction', 'input', 'output'],
        num_rows: 5
    })
})


In [29]:
print(dataset["train"][0])

{'instruction': 'Generate a concise analytical narrative from the structured findings. Use only the provided facts.', 'input': {'domain': 'sales', 'metrics': [{'name': 'Total Revenue', 'value': 1250000.0, 'unit': 'INR'}, {'name': 'Profit Margin', 'value': 19.39, 'unit': 'percent'}], 'trends': [{'metric': 'Revenue', 'direction': 'increasing', 'change': 12.4, 'unit': 'percent'}], 'comparisons': [{'dimension': 'category', 'metric': 'Revenue', 'highest': {'name': 'Electronics', 'value': 450000.0}, 'lowest': {'name': 'Furniture', 'value': 180000.0}}], 'anomalies': []}, 'output': 'Revenue reached INR 1,250,000 with a profit margin of 19.39%. Revenue increased by 12.4% during the reporting period. Electronics generated the highest category revenue at INR 450,000, while Furniture generated the lowest at INR 180,000.'}


In [30]:
def format_example(example):

    input_json = json.dumps(
        example["input"],
        indent=2,
        default=str
    )

    messages = [
        {
            "role": "user",
            "content": f"""{example["instruction"]}

Structured analytical findings:

{input_json}

Generate the analytical narrative now."""
        },
        {
            "role": "assistant",
            "content": example["output"]
        }
    ]

    return {
        "messages": messages
    }

In [31]:
formatted_dataset = dataset["train"].map(format_example)

print(formatted_dataset[0])

Map:   0%|          | 0/5 [00:00<?, ? examples/s]

{'instruction': 'Generate a concise analytical narrative from the structured findings. Use only the provided facts.', 'input': {'domain': 'sales', 'metrics': [{'name': 'Total Revenue', 'value': 1250000.0, 'unit': 'INR'}, {'name': 'Profit Margin', 'value': 19.39, 'unit': 'percent'}], 'trends': [{'metric': 'Revenue', 'direction': 'increasing', 'change': 12.4, 'unit': 'percent'}], 'comparisons': [{'dimension': 'category', 'metric': 'Revenue', 'highest': {'name': 'Electronics', 'value': 450000.0}, 'lowest': {'name': 'Furniture', 'value': 180000.0}}], 'anomalies': []}, 'output': 'Revenue reached INR 1,250,000 with a profit margin of 19.39%. Revenue increased by 12.4% during the reporting period. Electronics generated the highest category revenue at INR 450,000, while Furniture generated the lowest at INR 180,000.', 'messages': [{'content': 'Generate a concise analytical narrative from the structured findings. Use only the provided facts.\n\nStructured analytical findings:\n\n{\n  "domain": 

Load Qwen in 4-Bit Mode

In [32]:
from transformers import AutoTokenizer, AutoModelForCausalLM, BitsAndBytesConfig
import torch

model_name = "Qwen/Qwen2.5-1.5B-Instruct"

bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.float16,
    bnb_4bit_use_double_quant=True
)

tokenizer = AutoTokenizer.from_pretrained(model_name)

model = AutoModelForCausalLM.from_pretrained(
    model_name,
    quantization_config=bnb_config,
    device_map="auto"
)

print("Qwen loaded in 4-bit mode!")

Loading weights:   0%|          | 0/338 [00:00<?, ?it/s]

Qwen loaded in 4-bit mode!


In [33]:
from peft import prepare_model_for_kbit_training

model = prepare_model_for_kbit_training(model)

print("Model prepared for QLoRA training.")

Model prepared for QLoRA training.


In [40]:
from peft import LoraConfig
from trl import SFTTrainer, SFTConfig

peft_config = LoraConfig(
    r=16,
    lora_alpha=32,
    lora_dropout=0.05,
    bias="none",
    task_type="CAUSAL_LM",
    target_modules=[
        "q_proj",
        "k_proj",
        "v_proj",
        "o_proj",
        "gate_proj",
        "up_proj",
        "down_proj"
    ]
)

training_args = SFTConfig(
    output_dir="qwen_lora_output",
    num_train_epochs=3,
    per_device_train_batch_size=1,
    gradient_accumulation_steps=4,
    learning_rate=2e-4,
    logging_steps=1,
    save_strategy="epoch",
    fp16=False,
    bf16=True,
    report_to="none"
)

trainer = SFTTrainer(
    model=model,
    args=training_args,
    train_dataset=formatted_dataset,
    processing_class=tokenizer,
    peft_config=peft_config
)

trainer.train()

/usr/local/lib/python3.13/dist-packages/peft/mapping_func.py:148: UserWarning: You are trying to modify a model with PEFT for a second time. If you want to reload the model with a different config, make sure to call `.unload()` before.
  warnings.warn(
/usr/local/lib/python3.13/dist-packages/peft/tuners/tuners_utils.py:331: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(


Step,Training Loss
1,1.773160
2,1.429262
3,1.340245
4,0.970471
5,1.060549
6,0.796219


TrainOutput(global_step=6, training_loss=1.2283175786336262, metrics={'train_runtime': 30.1134, 'train_samples_per_second': 0.498, 'train_steps_per_second': 0.199, 'total_flos': 32624829591552.0, 'train_loss': 1.2283175786336262, 'epoch': 3.0})

In [41]:
# Save the trained LoRA adapter
trainer.save_model("qwen_lora_output")
tokenizer.save_pretrained("qwen_lora_output")

print("LoRA adapter saved successfully.")

LoRA adapter saved successfully.


In [42]:
from peft import PeftModel
from transformers import AutoModelForCausalLM, AutoTokenizer, BitsAndBytesConfig
import torch

model_name = "Qwen/Qwen2.5-1.5B-Instruct"

bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.float16,
    bnb_4bit_use_double_quant=True
)

tokenizer = AutoTokenizer.from_pretrained(model_name)

base_model = AutoModelForCausalLM.from_pretrained(
    model_name,
    quantization_config=bnb_config,
    device_map="auto"
)

finetuned_model = PeftModel.from_pretrained(
    base_model,
    "qwen_lora_output"
)

print("Fine-tuned QLoRA model loaded.")

Loading weights:   0%|          | 0/338 [00:00<?, ?it/s]

Fine-tuned QLoRA model loaded.


In [45]:
def generate_narrative(model, insights, max_new_tokens=300):

    insights_json = json.dumps(insights, indent=2)

    prompt = f"""
You are a business analytics assistant.

Generate a concise analytical narrative from the structured
analytical findings provided below.

Rules:
1. Use ONLY the information provided.
2. Do not invent numbers, facts, trends, or causes.
3. Mention the most important metrics.
4. Mention important trends.
5. Mention important comparisons.
6. Mention anomalies if they are provided.
7. Ignore empty sections.
8. Do not claim that one factor caused another unless causation
   is explicitly provided.
9. Write in a professional business-analysis style.

Structured analytical findings:
{insights_json}

Generate the business narrative now.
"""

    messages = [
        {"role": "user", "content": prompt}
    ]

    text = tokenizer.apply_chat_template(
        messages,
        tokenize=False,
        add_generation_prompt=True
    )

    inputs = tokenizer(text, return_tensors="pt").to(model.device)

    with torch.no_grad():
        outputs = model.generate(
            **inputs,
            max_new_tokens=max_new_tokens,
            do_sample=False
        )

    generated_tokens = outputs[0][inputs["input_ids"].shape[1]:]

    return tokenizer.decode(
        generated_tokens,
        skip_special_tokens=True
    )

In [46]:
finetuned_response = generate_narrative(
    finetuned_model,
    healthcare_insights
)

print("Fine-tuned Model Output:")
print(finetuned_response)

Fine-tuned Model Output:
The dataset contains 5000 records related to healthcare patients. The main metrics analyzed include the average wait time and the readmission rate. The average wait time for all patients was found to be 42.5 minutes with an average age of 35 years old. The department with the longest average wait time was Emergency (68.4 minutes), while Dermatology had the shortest average wait time at 21.7 minutes. The readmission rate among patients was 8.2%, indicating that approximately 1 out of every 12 patients were readmitted within 30 days after discharge.


In [47]:
baseline_response = response

print("Baseline response saved.")

Baseline response saved.


In [49]:
print("BASELINE:")
print(baseline_response)

print("\nFINE-TUNED:")
print(finetuned_response)

BASELINE:
The healthcare dataset contains records for 5,000 patients across various departments and diagnoses. The primary focus of this analysis is on two key metrics: the average wait time and the readmission rate.

The average wait time for all patients is 42.5 minutes, with an average readmission rate of 8.2 percent. These figures provide a baseline understanding of typical patient experiences within each department.

By examining the data further, we can identify specific areas where services may be more efficient or where improvements could be made. For instance, Emergency departments have the longest average wait times at 68.4 minutes compared to other departments like Dermatology, which has the shortest average wait time at 21.7 minutes.

These insights suggest that there might be opportunities for optimization in departments with longer wait times, such as Emergency, to reduce patient wait times and improve overall service efficiency. Additionally, focusing on reducing the rea

In [48]:
baseline_validation = validate_narrative(
    baseline_response,
    healthcare_insights
)

finetuned_validation = validate_narrative(
    finetuned_response,
    healthcare_insights
)

print("BASELINE VALIDATION")
print(baseline_validation)

print("\nFINE-TUNED VALIDATION")
print(finetuned_validation)

BASELINE VALIDATION
{'valid': True, 'missing_values': []}

FINE-TUNED VALIDATION
{'valid': True, 'missing_values': []}


In [50]:
evaluation = {
    "Model": ["Baseline", "QLoRA"],
    "Coverage": [None, None],
    "Unsupported_Claims": [None, None],
    "Factual_Consistency": [None, None],
    "Readability": [None, None]
}

import pandas as pd

evaluation_df = pd.DataFrame(evaluation)
evaluation_df

,Model,Coverage,Unsupported_Claims,Factual_Consistency,Readability
0,Baseline,None,None,None,None
1,QLoRA,None,None,None,None


In [51]:
# Facts that are explicitly present in healthcare_insights

supported_facts = {
    "5000",
    "42.5",
    "8.2",
    "68.4",
    "21.7",
    "Emergency",
    "Dermatology"
}

supported_facts

{'21.7', '42.5', '5000', '68.4', '8.2', 'Dermatology', 'Emergency'}

In [52]:
def calculate_coverage(narrative, supported_facts):
    narrative_lower = narrative.lower()

    found = []
    missing = []

    for fact in supported_facts:
        if fact.lower() in narrative_lower:
            found.append(fact)
        else:
            missing.append(fact)

    coverage = len(found) / len(supported_facts) * 100

    return coverage, found, missing

In [53]:
def calculate_coverage(narrative, supported_facts):
    narrative_lower = narrative.lower()

    found = []
    missing = []

    for fact in supported_facts:
        if fact.lower() in narrative_lower:
            found.append(fact)
        else:
            missing.append(fact)

    coverage = len(found) / len(supported_facts) * 100

    return coverage, found, missing

In [54]:
baseline_coverage = calculate_coverage(
    baseline_response,
    supported_facts
)

finetuned_coverage = calculate_coverage(
    finetuned_response,
    supported_facts
)

print("Baseline:")
print(baseline_coverage)

print("\nQLoRA:")
print(finetuned_coverage)

Baseline:
(85.71428571428571, ['68.4', '42.5', '21.7', 'Emergency', '8.2', 'Dermatology'], ['5000'])

QLoRA:
(100.0, ['68.4', '42.5', '21.7', 'Emergency', '5000', '8.2', 'Dermatology'], [])


In [55]:
import re

def extract_numbers(text):
    return set(
        re.findall(r'\b\d+(?:\.\d+)?\b', text)
    )

In [56]:
input_numbers = extract_numbers(
    json.dumps(healthcare_insights)
)

baseline_numbers = extract_numbers(baseline_response)
finetuned_numbers = extract_numbers(finetuned_response)

print("Input numbers:")
print(input_numbers)

print("\nBaseline numbers:")
print(baseline_numbers)

print("\nQLoRA numbers:")
print(finetuned_numbers)

Input numbers:
{'42.5', '21.7', '5000', '68.4', '8.2'}

Baseline numbers:
{'42.5', '21.7', '5', '68.4', '000', '8.2'}

QLoRA numbers:
{'42.5', '21.7', '30', '5000', '35', '68.4', '8.2', '12', '1'}


In [57]:
baseline_unsupported = baseline_numbers - input_numbers
finetuned_unsupported = finetuned_numbers - input_numbers

print("Baseline unsupported numbers:")
print(baseline_unsupported)

print("\nQLoRA unsupported numbers:")
print(finetuned_unsupported)

Baseline unsupported numbers:
{'5', '000'}

QLoRA unsupported numbers:
{'35', '12', '30', '1'}


In [58]:
def unsupported_claim_score(narrative, input_data):
    input_numbers = extract_numbers(
        json.dumps(input_data)
    )

    output_numbers = extract_numbers(narrative)

    unsupported = output_numbers - input_numbers

    return len(unsupported), unsupported

In [59]:
baseline_unsupported_count, baseline_unsupported = \
    unsupported_claim_score(
        baseline_response,
        healthcare_insights
    )

finetuned_unsupported_count, finetuned_unsupported = \
    unsupported_claim_score(
        finetuned_response,
        healthcare_insights
    )

print("Baseline:", baseline_unsupported_count, baseline_unsupported)
print("QLoRA:", finetuned_unsupported_count, finetuned_unsupported)

Baseline: 2 {'5', '000'}
QLoRA: 4 {'35', '12', '30', '1'}


In [60]:
evaluation_df.loc[0, "Coverage"] = baseline_coverage[0]
evaluation_df.loc[1, "Coverage"] = finetuned_coverage[0]

evaluation_df.loc[0, "Unsupported_Claims"] = baseline_unsupported_count
evaluation_df.loc[1, "Unsupported_Claims"] = finetuned_unsupported_count

evaluation_df

,Model,Coverage,Unsupported_Claims,Factual_Consistency,Readability
0,Baseline,85.714286,2,None,None
1,QLoRA,100.0,4,None,None


In [63]:
import re
import json

def extract_numbers(text):
    return set(re.findall(r'\b\d+(?:\.\d+)?\b', text))

def factual_consistency(narrative, insights):
    input_numbers = extract_numbers(json.dumps(insights))
    output_numbers = extract_numbers(narrative)

    unsupported_numbers = output_numbers - input_numbers

    if len(output_numbers) == 0:
        return 100.0

    supported_numbers = output_numbers - unsupported_numbers

    return (len(supported_numbers) / len(output_numbers)) * 100

In [64]:
baseline_factual = factual_consistency(
    baseline_response,
    healthcare_insights
)

qlora_factual = factual_consistency(
    finetuned_response,
    healthcare_insights
)

evaluation_df.loc[0, "Factual_Consistency"] = baseline_factual
evaluation_df.loc[1, "Factual_Consistency"] = qlora_factual

evaluation_df

,Model,Coverage,Unsupported_Claims,Factual_Consistency,Readability
0,Baseline,85.714286,2,66.666667,None
1,QLoRA,100.0,4,55.555556,None


In [65]:
!pip install textstat

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 177.1/177.1 kB 5.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.1/2.1 MB 31.3 MB/s eta 0:00:00


In [66]:
import textstat

def readability_score(narrative):
    return textstat.flesch_reading_ease(narrative)

baseline_readability = readability_score(baseline_response)
qlora_readability = readability_score(finetuned_response)

evaluation_df.loc[0, "Readability"] = baseline_readability
evaluation_df.loc[1, "Readability"] = qlora_readability

evaluation_df

,Model,Coverage,Unsupported_Claims,Factual_Consistency,Readability
0,Baseline,85.714286,2,66.666667,29.918235
1,QLoRA,100.0,4,55.555556,56.761103


Coverage: QLoRA is better — it included all supported facts, compared with 85.71% for the baseline.
Unsupported claims: Baseline has 2, while QLoRA has 4. This means the current QLoRA output contains more unsupported information.
Factual consistency: Baseline scores 66.67%, while QLoRA scores 55.56%. This is consistent with the unsupported numerical claims we already identified.
Readability: QLoRA scores 56.76 vs 29.92, so its generated narrative is substantially easier to read according to the Flesch Reading Ease metric.

In [67]:
def relevance_score(narrative, supported_facts):
    narrative_lower = narrative.lower()

    relevant_facts = [
        fact for fact in supported_facts
        if fact.lower() in narrative_lower
    ]

    if len(supported_facts) == 0:
        return 100.0

    return (len(relevant_facts) / len(supported_facts)) * 100

In [68]:
baseline_relevance = relevance_score(
    baseline_response,
    supported_facts
)

qlora_relevance = relevance_score(
    finetuned_response,
    supported_facts
)

evaluation_df.loc[0, "Relevance"] = baseline_relevance
evaluation_df.loc[1, "Relevance"] = qlora_relevance

evaluation_df

,Model,Coverage,Unsupported_Claims,Factual_Consistency,Readability,Relevance
0,Baseline,85.714286,2,66.666667,29.918235,85.714286
1,QLoRA,100.0,4,55.555556,56.761103,100.000000


In [69]:
def validate_narrative(narrative, insights):

    # Numbers present in the source insights
    input_numbers = extract_numbers(json.dumps(insights))

    # Numbers present in generated narrative
    output_numbers = extract_numbers(narrative)

    # Numbers that were not present in the source
    unsupported_numbers = output_numbers - input_numbers

    # Check if all source numbers are covered
    missing_numbers = input_numbers - output_numbers

    # Narrative is valid only if there are no unsupported numbers
    is_valid = len(unsupported_numbers) == 0

    return {
        "is_valid": is_valid,
        "unsupported_numbers": sorted(unsupported_numbers),
        "missing_numbers": sorted(missing_numbers)
    }

In [70]:
baseline_validation = validate_narrative(
    baseline_response,
    healthcare_insights
)

baseline_validation

{'is_valid': False,
 'unsupported_numbers': ['000', '5'],
 'missing_numbers': ['5000']}

In [71]:
qlora_validation = validate_narrative(
    finetuned_response,
    healthcare_insights
)

qlora_validation

{'is_valid': False,
 'unsupported_numbers': ['1', '12', '30', '35'],
 'missing_numbers': []}

In [72]:
evaluation_df["Validation"] = [
    baseline_validation["is_valid"],
    qlora_validation["is_valid"]
]

evaluation_df

,Model,Coverage,Unsupported_Claims,Factual_Consistency,Readability,Relevance,Validation
0,Baseline,85.714286,2,66.666667,29.918235,85.714286,False
1,QLoRA,100.0,4,55.555556,56.761103,100.000000,False


Coverage: QLoRA improved from 85.71 → 100%, meaning it covered more of the expected information.
Unsupported claims: Increased from 2 → 4, which means QLoRA introduced more claims that were not sufficiently supported.
Factual consistency: Decreased from 66.67 → 55.56%, suggesting the QLoRA output was less factually consistent according to this metric.
Readability: Improved substantially from 29.92 → 56.76.
Relevance: Improved from 85.71 → 100%.
Validation: Both are marked False, so neither output passed the validation criterion.

In [ ]:
OUTPUT_DIR = "./outputs/qlora_adapter"

trainer.save_model(OUTPUT_DIR)
tokenizer.save_pretrained(OUTPUT_DIR)

print(f"QLoRA adapter saved to: {OUTPUT_DIR}")